packages/@aws-cdk/aws-servicediscovery/lib/namespace.ts

```diff
@@ -0,0 +1,106 @@
+import cdk = require('@aws-cdk/cdk');
+
+export interface INamespace extends cdk.IConstruct {
+  /**
+   * A name for the Namespace.
+   */
+  readonly namespaceName: string;
+
+  /**
+   * Namespace Id for the Namespace.
+   */
+  readonly namespaceId: string;
+
+  /**
+   * Namespace ARN for the Namespace.
+   */
+  readonly namespaceArn: string;
+
+  /**
+   * Type of Namespace. Valid values: HTTP, DNS_PUBLIC, or DNS_PRIVATE
+   */
+  readonly type: NamespaceType;
+}
+
+export interface BaseNamespaceProps {
+  /**
+   * A name for the Namespace.
+   */
+  name: string;
+
+  /**
+   * A description of the Namespace.
+   *
+   * @default none
+   */
+  description?: string;
+}
+
+export interface NamespaceImportProps {
+  /**
+   * A name for the Namespace.
+   */
+  readonly namespaceName: string;
+
+  /**
+   * Namespace Id for the Namespace.
+   */
+  readonly namespaceId: string;
+
+  /**
+   * Namespace ARN for the Namespace.
+   */
+  readonly namespaceArn: string;
+
+  /**
+   * Type of Namespace. Valid values: HTTP, DNS_PUBLIC, or DNS_PRIVATE
+   */
+  readonly type: NamespaceType;
+}
+
+export enum NamespaceType {
+  /**
+   * Choose this option if you want your application to use only API calls to discover registered instances.
+   */
+  Http = "HTTP",
+
+  /**
+   * Choose this option if you want your application to be able to discover instances using either API calls or using
+   * DNS queries in a VPC.
+   */
+  DnsPrivate = "DNS_PRIVATE",
+
+  /**
+   * Choose this option if you want your application to be able to discover instances using either API calls or using
+   * public DNS queries. You aren't required to use both methods.
+   */
+  DnsPublic = "DNS_PUBLIC",
+}
+
+export abstract class NamespaceBase extends cdk.Construct implements INamespace {
+  public abstract readonly namespaceId: string;
+  public abstract readonly namespaceArn: string;
+  public abstract readonly namespaceName: string;
+  public abstract readonly type: NamespaceType;
+}
+
+export class ImportedNamespace extends NamespaceBase {
```

Comment:
Don't export the `ImportedFoo` classes.

Reply:
When i take away the "export", compiler complains that it's never used. Looking at the guidelines, I'm wondering if I need to have a separate Imported*Namespace for each of the three kinds of namespaces?

Reply:
It shouldn't be "unused" if you instantiate this from an `import` method...

Reply:
Did you make the 

```ts
class Namespace {
  public static import(...) { }
}
```

function?

Reply:
And yes, I think you're going to need 3 imported classes, otherwise there is no place to put the type-specific `createService()` methods. So, naively I would expect the following:

```ts
export interface INamespace { ... }

export interface IDnsNamespace extends INamespace {
  createService(...: DnsServiceProps): IService;
}

export class DnsNamespace implements IDnsNamespace {
  public static import(...): IDnsNamespace {
    return new ImportedDnsNamespace(...);
  }
}

class ImportedDnsNamespace implements IDnsNamespace {
}
```

Now, you can say you don't want to do that right now because it's going to take too much time, which I think by now is fair enough. If you don't, please make a ticket for it.

Reply:
Punting to https://github.com/awslabs/aws-cdk/issues/1996